docs/specs/m2m.md

```diff
@@ -0,0 +1,1152 @@
+  updateResource(input: UpdateResourceInput!): UpdateResourcePayload!
+  deleteResource(input: DeleteResourceInput!): DeleteResourcePayload!
+
+  createScope(input: CreateScopeInput!): CreateScopePayload!
+  updateScope(input: UpdateScopeInput!): UpdateScopePayload!
+  deleteScope(input: DeleteScopeInput!): DeleteScopePayload!
+
+  addResourceToClientID(input: AddResourceToClientIDInput!): AddResourceToClientIDPayload!
+  removeResourceFromClientID(input: RemoveResourceFromClientIDInput!): RemoveResourceFromClientIDPayload!
+  addScopeToClientID(input: AddScopeToClientIDInput!): AddScopeToClientIDPayload!
+  removeScopeFromClientID(input: RemoveScopeFromClientIDInput!): RemoveScopeFromClientIDPayload!
+}
+
+type Resource implements Entity & Node {
+  id: ID!
+  createdAt: DateTime!
+  updatedAt: DateTime!
+  uri: String!
+  name: String
+  """If clientID is null, then all scopes of this Resource is returned."""
+  """If clientID is specified, then only scopes that are associated with clientID is returned."""
+  """If searchKeyword is non-null, a prefix search of scope is performed."""
+  """If both clientID and searchKeyword are specified, they are AND-ed."""
+  scopes(clientID: String, searchKeyword: String, after: String, before: string, first: Int, last: Int): ScopeConnection
+  """The list of client IDs associated with this Resource."""
+  clientIDs: [String!]!
+}
+
+type Scope implements Entity & Node {
+  id: ID!
+  createdAt: DateTime!
+  updatedAt: DateTime!
+  resource: Resource!
+  scope: String!
+  description: String
+}
+
+type ResourceEdge {
+  cursor: String!
+  resource: Resource
+}
+
+type ResourceConnection {
+  edges: [ResourceEdge]
+  pageInfo: PageInfo!
+  totalCount: Int
+}
+
+type ScopeEdge {
+  cursor: String!
+  scope: Scope
+}
+
+type ScopeConnection {
+  edges: [ScopeEdge]
+  pageInfo: PageInfo!
+  totalCount: Int
+}
+
+input CreateResourceInput {
+  uri: String!
+  name: String
+}
+
+type CreateResourcePayload {
+  resource: Resource!
+}
+
+input UpdateResourceInput {
+  id: ID!
+  """The new name"""
+  name: String
+}
+
+type UpdateResourcePayload {
+  resource: Resource!
+}
+
+input DeleteResourceInput {
+  id: ID!
+}
+
+type DeleteResourcePayload {
+  ok: Boolean
+}
+
+input CreateScopeInput {
+  resourceID: ID!
+  scope: String!
+  description: String
+}
+
+type CreateScopePayload {
+  scope: Scope!
+}
+
+input UpdateScopeInput {
+  id: ID!
+  """The new description"""
+  description: String
+}
+
+type UpdateScopePayload {
+  scope: Scope!
+}
+
+input DeleteScopeInput {
+  id: ID!
+}
+
+type DeleteScopePayload {
+  ok: Boolean
+}
+
+input AddResourceToClientIDInput {
+  resourceID: ID!
+  clientID: String!
+}
+
+type AddResourceToClientIDPayload {
+  resource: Resource!
+}
+
+input RemoveResourceFromClientIDInput {
+  resourceID: ID!
+  clientID: String!
+}
+
+type RemoveResourceFromClientIDPayload {
+  resource: Resource!
+}
+
+input AddScopeToClientIDInput {
+  scopeID: ID!
```

Comment:
I think we should use `scope` to reference the scope here instead of ID, like how we use a key to reference a role. Therefore I suggest the input should be:

```
input AddScopeToClientIDInput {
  resourceUri: String!
  scope: String!
  clientID: String!
}
```

Reply:
Agree. But `resourceUri` should be `resourceURI`

Reply:
Please update the spec to reflect this proposal.